src/inc/til/env.h

```diff
@@ -0,0 +1,515 @@
+                            }
+                            valueName.resize(valueNameSize);
+                            valueData.resize(valueDataSize);
+
+                            if (valueNameSize)
+                            {
+                                std::wstring data;
+                                if (pass == 0 && (type == REG_SZ) && valueDataSize >= sizeof(wchar_t))
+                                {
+                                    data = {
+                                        reinterpret_cast<wchar_t*>(valueData.data()), valueData.size() / sizeof(wchar_t)
+                                    };
+                                }
+                                else if (pass == 1 && (type == REG_EXPAND_SZ) && valueDataSize >= sizeof(wchar_t))
+                                {
+                                    data = {
+                                        reinterpret_cast<wchar_t*>(valueData.data()), valueData.size() / sizeof(wchar_t)
+                                    };
+                                    data = expand_environment_strings(data.data());
+                                }
+
+                                // Because Registry data may or may not be null terminated... check if we've managed
+                                // to store an extra null in the wstring by telling it to create itself from pointer and size.
+                                // If we did, pull it off.
+                                if (!data.empty())
+                                {
+                                    if (data.back() == L'\0')
+                                    {
+                                        data = data.substr(0, data.size() - 1);
+                                    }
+                                }
+
+                                if (!data.empty())
+                                {
+                                    if (is_path_var(valueName))
+                                    {
+                                        concat_var(valueName, std::wstring{ data });
+                                    }
+                                    else
+                                    {
+                                        set_user_environment_var(valueName, std::wstring{ data });
+                                    }
+                                }
+                            }
+
+                            valueName.resize(maxValueNameSize);
+                            valueData.resize(maxValueDataSize);
+                            valueNameSize = maxValueNameSize;
+                            valueDataSize = maxValueDataSize;
+                            index++;
+                        }
+                        index = 0;
+                    }
+                }
+            }
+        }
+
+        std::wstring expand_environment_strings(std::wstring input)
+        {
+            // TODO: this should be replacing from ourselves, not from the OS

```

Comment:
See discussion here about how this: https://github.com/microsoft/terminal/pull/12516#discussion_r1104167687

Reply:
As #12516 is now closed, copying the discussion here.

I said:
>I take it this method needs implementing as follows - or is this too naive?
>
>Iterate through the string - append characters to a "return" buffer until you find a percent character. At this point start appending the characters to a "key" buffer instead. When you find a "closing" percent, look the key up in the map. If found, append the looked up value to the return buffer, otherwise append the key as was (including the percent characters), then reset the state. If at the end there's stuff left in the key buffer, append it to the return buffer (including a literal percent character at the beginning).
>
>Test cases:
>
>1. "Foo%ENV%Baz" with an environment variable %ENV% defined as "Bar" would result in the string "FooBarBaz".
>2. "Foo%ENV%Baz" with no %ENV% variable would result in the string "Foo%ENV%Baz".
>3. "Foo%ENV" would result in the string "Foo%ENV".

@KalleOlaviNiemitalo said:
>How about "Foo%Baz%ENV%" with an %ENV% variable but no %Baz% variable? CMD seems to expand %ENV% in that situation, but I don't know what ExpandEnvironmentStringsW does.

I can confirm that `ExpandEnvironmentStringsW` does indeed expand out %ENV% in this case - with it set to "ThisIsEnv", the following test passes:

```cpp
const auto result = wil::ExpandEnvironmentStringsW<std::wstring, 256>(L"Foo%Baz%ENV%");
VERIFY_ARE_EQUAL(L"Foo%BazThisIsEnv", result);
```

That _may_ suggest that `ExpandEnvironmentStringsW` has the (undocumented) behaviour that if an environment variable doesn't match, the "closing" percent can become the beginning of the next "key".

Reply:
I'm alright with your proposed implementation, even if it doesn't address this corner case, fwiw.